### PR TITLE
[PLAT-8778] EAS sourcemaps plugin monorepo compatibility

### PR DESCRIPTION
**Pull request by gingerbenw**
## Goal

To enable source-maps plugin to work for an expo project in a monorepo, based on yarn workspaces

## Design

The `@bugsnag/expo-eas-sourcemaps-plugin` updates the android and iOS build configuration with the necessary steps to execute an upload using the `@bugsnag/source-maps` library, and has a hardcoded path to the relevant binaries. (in node_modules)

We must prevent these binaries from being hoisted when using yarn workspaces, so after selecting the option to automatically uploads source maps, the cli tool will:

- Install necessary dev dependencies
- Update `app.json` (expo config) with the new plugin
- Detect version of yarn being used (either classic or version 2 and above)
- Detect whether using yarn workspaces
- Update `package.json` based on the yarn version and workspaces configuration

This will prevent the dependencies from being hoisted, and therefore enable the plugin to upload source maps during `eas build`

## Testing

Installed and built expo project using eas from within a monorepo, and as a standalone application using yarn v1.x and v3.x, tests updated

### CODE REVIEW

File: packages/expo-cli/lib/configure-plugin.js [hunk 1]
```diff
@@ -24,4 +24,28 @@ module.exports = async (projectRoot) => {
...
+    const packageJsonPath = join(projectRoot, 'package.json')
+    const packageJson = JSON.parse(await promisify(readFile)(packageJsonPath))
+    packageJson.workspaces = packageJson.workspaces || {}
+    packageJson.workspaces.nohoist = packageJson.workspaces.nohoist || []
+    if (!packageJson.workspaces.nohoist.includes(plugin)) {
+      packageJson.workspaces.nohoist.push(plugin)
+    }
+    if (!packageJson.workspaces.nohoist.includes(sourceMaps)) {
+      packageJson.workspaces.nohoist.push(sourceMaps)
+    }
```
Comment (kattrali) [inline packages/expo-cli/lib/configure-plugin.js]: Does this make sense all the time, or only when `workspaces` is already defined? Would creating a `workspaces` entry have other consequences for users' projects?

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: We do need to add the `workspaces` parameter if it's not already there, and outside of a yarn workspaces repository, will have no effect. There may well be some way to detect whether we're using yarn workspaces before we add the missing parameter though?

Comment (djskinner) [reply packages/expo-cli/lib/configure-plugin.js]: if I were using this I would prefer it not to add unnecessary stuff to my package.json (if I'm not using yarn workspaces). It looks like the workspaces entry is required by definition for workspaces to be enabled. See https://classic.yarnpkg.com/lang/en/docs/workspaces/#toc-how-to-use-it - so perhaps could test with `packageJson.workspaces && packageJson.private` and not add this stuff otherwise?

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: the workspaces entry needs to be present in the **root** package.json, but not the specific package. so we can't expect it to be there. but I agree that installing it when you don't need it isn't the best. will reconsider

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: Updated to detect yarn workspaces before adding the nohoist configuration

File: packages/expo-cli/lib/configure-plugin.js [hunk 2]
```diff
@@ -1,9 +1,32 @@
...
+    proc.stderr.on('data', d => {})
+
+    proc.on('error', err => { reject(err) })
+
+    proc.on('close', code => {
+      if (code === 0) {
+        resolve(true)
+      }
+
+      resolve(false)
```
Comment (kattrali) [inline packages/expo-cli/lib/configure-plugin.js]:
```suggestion
      resolve(code === 0)
```

File: packages/plugin-expo-eas-sourcemaps/package.json [hunk 3]
```diff
@@ -19,7 +19,7 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1"
   },
-  "peerDependencies": {
+  "dependencies": {
```
Comment (kattrali) [inline packages/plugin-expo-eas-sourcemaps/package.json]: Why is this needed?

File: packages/expo-cli/lib/install-plugin.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 const { spawn } = require('child_process')
 
 function resolveCommand (options) {
-  const command = ['install', '@bugsnag/plugin-expo-eas-sourcemaps']
+  const command = ['install', '@bugsnag/plugin-expo-eas-sourcemaps', '@bugsnag/source-maps']
```
Comment (kattrali) [inline packages/expo-cli/lib/install-plugin.js:4]: Is this needed if `@bugsnag/source-maps` is a direct dependency? (resolution here kinda depends on the answer to the other comment)

File: packages/plugin-expo-eas-sourcemaps/package.json (hunk 3 above)
Comment (gingerbenw) [reply packages/plugin-expo-eas-sourcemaps/package.json]: sorry, left over from testing. should be `peerDependency`

File: packages/expo-cli/lib/configure-plugin.js (hunk 2 above)
Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: derp. much cleaner. thanks

**Review by kattrali — APPROVED** [review]

File: packages/expo-cli/lib/configure-plugin.js [hunk 5]
```diff
@@ -1,9 +1,28 @@
...
 const { readFile, writeFile } = require('fs')
 const { promisify } = require('util')
+const { spawn } = require('child_process')
+const { blue } = require('kleur')
 
 const plugin = '@bugsnag/plugin-expo-eas-sourcemaps'
 
+function usingWorkspaces (projectRoot) {
+  return new Promise((resolve, reject) => {
+    const command = ['workspaces', 'info']
```
Comment (djskinner) [inline packages/expo-cli/lib/configure-plugin.js]: When I run `yarn workspaces info` on `yarn 3.2.2` I get command not found:

```
Unknown Syntax Error: Command not found; did you mean:

$ yarn workspaces list [--since] [-R,--recursive] [-v,--verbose] [--json]
While running workspaces info
```

Maybe this is because it's not a workspace but then it proceeds to list available commands (just list), which makes me think it would be listed if it exists in this version of yarn. Even after `yarn plugin import workspace-tools` I don't see the info command:

```
Unknown Syntax Error: Command not found; did you mean one of:

  0. yarn workspaces list [--since] [-R,--recursive] [-v,--verbose] [--json]
  1. yarn workspaces focus [--json] [--production] [-A,--all] ...
  2. yarn workspaces foreach [-R,--recursive] [--from #0] [-A,--all] [-v,--verbose] [-p,--parallel] [-i,--interlaced] [-j,--jobs #0] [-t,--topological] [--topological-dev] [--include #0] [--exclude #0] [--no-private] [--since] <commandName> ...

While running workspaces info
```

I can't find something definitive saying it was removed in yarn 2+ but it's not listed as a command here https://yarnpkg.com/cli/workspaces/list. 

So I'm wondering if this approach will only work for yarn 1.x?

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: ahh. that could be possible. I'll check if list covers what we're after

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: `yarn workspaces list`

```
yarn workspaces v1.22.18
error Invalid subcommand. Try "info, run"
info Visit https://yarnpkg.com/en/docs/cli/workspaces for documentation about this command.
```

opposite problem on yarn 1.x!

Comment (djskinner) [reply packages/expo-cli/lib/configure-plugin.js]: Also `yarn workspaces list --json`

gives me 

```
yarn workspaces list --json
{"location":".","name":"dashboard"}
```
even though the repo I ran it is is not a workspace.

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: how useful! okay, I'll have to consider another option - thanks!

**Comment by gingerbenw** [comment]
Now checks for version of yarn before running the appropriate commands to detect whether yarn workspaces is in use

**Review by djskinner — APPROVED** [review]

**Review by kattrali — APPROVED** [review]

File: packages/expo-cli/lib/configure-plugin.js [hunk 6]
```diff
@@ -1,9 +1,38 @@
 const { join } = require('path')
 const { readFile, writeFile } = require('fs')
 const { promisify } = require('util')
+const { spawn } = require('child_process')
+const { blue } = require('kleur')
 
 const plugin = '@bugsnag/plugin-expo-eas-sourcemaps'
 
+function usingYarnV2 (projectRoot) {
```
Comment (gareththackeray) [inline packages/expo-cli/lib/configure-plugin.js]: Not really a big deal but I guess this should be `usingYarnV2AndUp` or something?  At least when we do the manual documentation we should use that kind of phrasing anyway.

Comment (gingerbenw) [reply packages/expo-cli/lib/configure-plugin.js]: good shout - or I could flip it on it's head and make it `usingYarnClassic`

**Comment by gingerbenw** [comment]
Removed `installConfig.hoistingLimits` as:

- [this config option should be set in yarnrc.yml](https://yarnpkg.com/getting-started/migration#if-required-enable-the-node-modules-plugin)
- users likely won't get to the point of installing bugsnag without having already disabled PnP when using yarn 2, as it's not compatible with expo

covered in [manual install guide PR](https://github.com/bugsnag/docs.bugsnag.com/pull/1775)

**Comment by tomlongridge** [comment]
> Removed `installConfig.hoistingLimits` as:
> 
> * [there are multiple ways to install this plugin](https://yarnpkg.com/getting-started/migration#if-required-enable-the-node-modules-plugin)
> * users won't get to this point without having already disabled PnP when using yarn 2, as it's not compatible with expo
> 
> covered in [manual install guide PR](https://github.com/bugsnag/docs.bugsnag.com/pull/1775)

@gareththackeray – thoughts on this? It could be confusing to put the option in `package.json` if they don't need it, but on the other hand it might be better to err on having it in there rather than not at all. Documenting this for people using the CLI (i.e. not in the manual config page) isn't going to be straightforward as I can't see an obvious place.

**Comment by gareththackeray** [comment]
I spoke to @gingerbenw.  He's confident.  He's rephrasing it all to be clear.  So I'm happy.

**Comment by gingerbenw** [comment]
Turns out I had rephrased this badly, my apologies. 🙇🏻 

- `installConfig.hoistingLimits: "workspaces"` is the equivalent of using "no hoist", and has been re-added to the CLI configuration step
- node modules plugin will not be documented with our plugin installation as it's required for yarn 2+ and considered out of the scope of our plugin configuration
- making plugin compatible with hoisting to be considered in the roadmap
